File: algorithms/199.cpp
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution
{
private:
    void dfs(TreeNode *root, vector<int> &result, int h)
    {
        if ( h > (int)result.size() )
            result.push_back(root->val);
        if ( root->right )
            dfs(root->right, result, h+1);
        if ( root->left )
            dfs(root->left, result, h+1);
    }
public:
    vector<int> rightSideView(TreeNode* root)
    {
        vector<int> result;
        if ( root == NULL )
            return result;
        dfs(root, result, 1);
        return result;
    }
};

int main()
{
    return 0;
}

File: algorithms/python3/199.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        rls = []
        if not root:
            return rls
        self.helper(root, 0, rls)
        return rls

    def helper(self, root, depth, rls):
        if depth >= len(rls):
            rls.append(root.val)
        else:
            rls[depth] = root.val
        if root.left:
            self.helper(root.left, depth+1, rls)
        if root.right:
            self.helper(root.right, depth+1, rls)
